### PR TITLE
[MINI-6188] Send analytics to host app

### PR DESCRIPTION
**Pull request by GondiTilak**
# Description
[MINI-6188]:
As a host app partner MiniApp provider, I want to be able to send analytics to the host app so the host app can retrieve analytics for the MiniApp.
[MINI-6191]:
Implement interface for analytics and store the analytics calls in a text file. Create a new screen called from a button where testers will be able to read the file to see all the analytics calls they performed.

This PR contains the SDK Changes and sample app changes, for send analytics to host app:
- implementing new interface for receiving the MAAnalytics info object from MiniApps.
- This new interface didReceiveMAAnalytics(analyticsInfo:, completionHandler: ) must be implemented in the host app.
- Host app should implement interface/protocol to receive the analytics
- Host app can write the analytics in a text file which will be useful for QA to verify.
- This can be viewed under Settings-> QA -> Analytics Info-> Any view that shows the Analytics log text file

## Links
[MINI-6188](https://jira.rakuten-it.com/jira/browse/MINI-6188)
[MINI-6191](https://jira.rakuten-it.com/jira/browse/MINI-6191)

# Checklist
- [x] I have read the [contributing guidelines](CONTRIBUTING.md).
- [x] I have completed the [SDK Development Learning Path](CONTRIBUTING.md#SDK-Development-Learning-Path) (if submitting a feature PR)
- [x] I wrote/updated tests for new/changed code
- [ ] I removed all sensitive data before every commit, including API endpoints and keys
- [ ] I ran `fastlane ci` without errors

### CODE REVIEW

File: Sources/Classes/core/Models/MAAnalytics.swift [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+public class MAAnalytics: Codable {
+    let eventType: MAAnalyticsEventType
+    let actionType: MAAnalyticsActionType
+    let pageName: Int
+    let componentName: String
+    let elementType: String
+    let data: String
+
+    public init(eventType: MAAnalyticsEventType, actionType: MAAnalyticsActionType, pageName: Int, componentName: String, elementType: String, data: String) {
```
Comment (rleojoseph) [inline Sources/Classes/core/Models/MAAnalytics.swift:26]: Can we lint the constructor here ?

File: Sources/Classes/core/Models/MAAnalytics.swift [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    case close
+    case add
+    case delete
+    case change
+}
+
+public class MAAnalytics: Codable {
+    let eventType: MAAnalyticsEventType
+    let actionType: MAAnalyticsActionType
+    let pageName: Int
```
Comment (rleojoseph) [inline Sources/Classes/core/Models/MAAnalytics.swift:21]: pageName will not be a int

File: Sources/Classes/core/Models/MAAnalytics.swift [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+public enum MAAnalyticsActionType: String, Codable {
+    case open
+    case close
+    case add
+    case delete
+    case change
+}
+
+public class MAAnalytics: Codable {
```
Comment (rleojoseph) [inline Sources/Classes/core/Models/MAAnalytics.swift:18]:
```suggestion
public class MAAnalyticsInfo: Codable {
```

File: Sources/Classes/core/Protocols/MAAnalyticsDelegate.swift [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import Foundation
+
+/**
+ Public Protocol that will be used by the Mini App to send Aalytics info to host app.
```
Comment (rleojoseph) [inline Sources/Classes/core/Protocols/MAAnalyticsDelegate.swift:4]:
```suggestion
 Public Protocol that will be used by the Mini App to send Analytics info to host app.
```

File: USERGUIDE.md [hunk 5]
```diff
@@ -1148,6 +1148,23 @@ extension ViewController: MiniAppMessageDelegate {
 }
 ```
 
+<a id="miniapp-sendAnalytics"></a>
+
+### Receive MiniApp Analytics 
+
+MiniApps can send the analytics information to host app through the `sendAnalytics` interface defined in JS-SDK.
+MiniAppSDK in iOS provides the `MiniAppMessageDelegate` interface method `func didReceiveMAAnalytics(analyticsInfo: MAAnalytics, completionHandler: @escaping (Result<MASDKProtocolResponse, MAAnalyticsError>) -> Void)` which can be implemented in the host app , to receive the MiniApp analytics event info object of type `MAAnalytics`. This delegate method is an optional.
```
Comment (rleojoseph) [inline USERGUIDE.md:1156]:
```suggestion
MiniApps can send the analytics information to host app through the `sendAnalytics` interface defined in JS-SDK.
iOS Hostapp can extend this optional 'MAAnalayticsDelegate' interface/delegate to receive the analytics that is sent from the Miniapp. 
```

**Review by rleojoseph — COMMENTED** [review]
Just few comments. Take a look

**Review by Climbatize — APPROVED** [review]
LGTM, nice job guys!

File: Sources/Classes/core/Protocols/MAAnalyticsDelegate.swift [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import Foundation
+
+/**
+ Public Protocol that will be used by the Mini App to send Analytics info to host app.
+ With the Native implementation host can receive MAAnalyticsInfo object through this interface.
+ */
+public protocol MAAnalyticsDelegate: AnyObject {
+    /// This interface must be implemented int he host app to get the MAAnalyticsInfo object from the MiniApps.
```
Comment (rleojoseph) [inline Sources/Classes/core/Protocols/MAAnalyticsDelegate.swift:8]:
```suggestion
    /// This interface must be implemented in the host app to get the MAAnalyticsInfo object from the MiniApps.
```

File: USERGUIDE.md [hunk 7]
```diff
@@ -1148,6 +1148,23 @@ extension ViewController: MiniAppMessageDelegate {
...
+<a id="miniapp-sendAnalytics"></a>
+
+### Receive MiniApp Analytics 
+
+MiniApps can send the analytics information to host app through the `sendAnalytics` interface defined in JS-SDK.
+iOS Hostapp can extend this optional `MAAnalayticsDelegate` interface/delegate to receive the analytics that is sent from the Miniapp.
+
+```swift
+extension ViewController: MiniAppMessageDelegate {
+    func didReceiveMAAnalytics(analyticsInfo: MAAnalytics, completionHandler: @escaping (Result<MASDKProtocolResponse, MAAnalyticsError>) -> Void) {
```
Comment (rleojoseph) [inline USERGUIDE.md:1160]:
```suggestion
    func didReceiveMAAnalytics(analyticsInfo: MAAnalyticsInfo, completionHandler: @escaping (Result<MASDKProtocolResponse, MAAnalyticsError>) -> Void) {
```

**Review by rleojoseph — COMMENTED** [review]
Overall looks good to me. Just few changes, take a look

File: CHANGELOG.md [hunk 8]
```diff
@@ -7,6 +7,7 @@
 - **Feature:** Added new `HostAppThemeColors` model class to hold the primary and secondary color strings in Hex format.
 - **Feature:** Added a new optional interface `getHostAppThemeColors(completionHandler:)` in `MiniAppMessageDelegate` protocol, to fetch the host app theme colors.
 - **Feature:** Added new `isDarkMode` interface to get the mode set in Host app / Device.
+- **Feature:** Added new public interface `didReceiveMAAnalytics(analyticsInfo:, completionHandler: @escaping (Result<MASDKProtocolResponse, MAAnalyticsError>) -> Void)` for host app to receive the `MAAnalyticsInfo` info for the events from MiniApps.
```
[the remark this answers is not in the document]
Comment (GondiTilak) [reply CHANGELOG.md]: Thank you, I have updated this 👍🏽 please review.

File: Example/Controllers/SwiftUI/Features/Settings/Features/MAAnalyticsInfoView.swift [hunk 9]
```diff
@@ -0,0 +1,35 @@
+import SwiftUI
+
+struct MAAnalyticsInfoView: View {
+    @State var analyticsData: String = "No Data Found"
```
[the remark this answers is not in the document]
Comment (GondiTilak) [reply Example/Controllers/SwiftUI/Features/Settings/Features/MAAnalyticsInfoView.swift]: Thank you, I have updated this 👍🏽 please review.

File: Example/Controllers/SwiftUI/Helper/MAAnalyticsInfoLogger.swift [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+        let fileName = "MAAnalyticsInfoLogs.txt"
+        return documentsDirectory.appendingPathComponent(fileName)
+    }
+
+    static func startLogs(forMiniApp miniAppId: String) {
+        MAAnalyticsInfoLogger.logAnalytics("--------------------------------- \nMiniApp -> \(miniAppId) is OPEN")
+    }
+
+    func stopLogs(forMiniApp miniAppId: String) {
+        MAAnalyticsInfoLogger.logAnalytics("MiniApp -> \(miniAppId) is CLOSE \n ---------------------------------")
```
[the remark this answers is not in the document]
Comment (GondiTilak) [reply Example/Controllers/SwiftUI/Helper/MAAnalyticsInfoLogger.swift]: Thank you, I have updated this 👍🏽 please review.

File: Sources/Classes/core/JavascriptBridge/MiniAppScriptMessageHandler.swift [hunk 11]
```diff
@@ -829,6 +831,24 @@ extension MiniAppScriptMessageHandler {
     }
 }
 
+//MARK: MiniApp Receive MAAnlytics Info
+extension MiniAppScriptMessageHandler {
+    func didReceiveMAAnalytics(requestParam: RequestParameters?, callbackId: String){
+        guard let requestParamValue = requestParam?.analyticsInfo else {
+            executeJavaScriptCallback(responseStatus: .onError, messageId: callbackId, response: prepareMAJavascriptError(MiniAppJavaScriptError.valueIsEmpty))
```
[the remark this answers is not in the document]
Comment (GondiTilak) [reply Sources/Classes/core/JavascriptBridge/MiniAppScriptMessageHandler.swift]: Thank you, I have updated this 👍🏽 please review.

File: Sources/Classes/core/Protocols/MAAnalyticsDelegate.swift (hunk 6 above)
Comment (GondiTilak) [reply Sources/Classes/core/Protocols/MAAnalyticsDelegate.swift]: Thank you, I have updated this 👍🏽 please review.

File: USERGUIDE.md (hunk 7 above)
Comment (GondiTilak) [reply USERGUIDE.md]: Thank you, I have updated this 👍🏽 please review.

File: USERGUIDE.md [hunk 12]
```diff
@@ -1148,6 +1148,23 @@ extension ViewController: MiniAppMessageDelegate {
...
+        print(analyticsInfo)
+        // analyticsInfo: MAAnalytics - This object holds the Analytics info of the events triggered from MiniApps, This can be further used to record the MiniApp Analytics events from host app.
```
[the remark this answers is not in the document]
Comment (GondiTilak) [reply USERGUIDE.md]: Thank you, I have updated this 👍🏽 please review.

**Comment by GondiTilak** [comment]
> LGTM, nice job guys!

Thank you 👍🏽

File: .swiftlint.yml [hunk 13]
```diff
@@ -2,6 +2,7 @@ disabled_rules: # rule identifiers to exclude from running
   - nesting
   - orphaned_doc_comment
   - superfluous_disable_command
+  - unneeded_synthesized_initializer
```
Comment (rleojoseph) [inline .swiftlint.yml:5]: We don't need it to declare for global as of now.

File: Sources/Classes/core/JavascriptBridge/MiniAppScriptMessageHandler.swift [hunk 14]
```diff
@@ -829,6 +831,24 @@ extension MiniAppScriptMessageHandler {
     }
 }
 
+//MARK: MiniApp Receive MAAnlytics Info
```
Comment (rleojoseph) [inline Sources/Classes/core/JavascriptBridge/MiniAppScriptMessageHandler.swift:834]:
```suggestion
//MARK: MiniApp Receive MAAnalyticsInfo
```

File: USERGUIDE.md [hunk 15]
```diff
@@ -1148,6 +1148,23 @@ extension ViewController: MiniAppMessageDelegate {
...
 
+<a id="miniapp-sendAnalytics"></a>
+
+### Receive MiniApp Analytics 
+
+MiniApps can send the analytics information to host app through the `sendAnalytics` interface defined in JS-SDK.
+iOS Hostapp can extend this optional `MAAnalayticsDelegate` interface/delegate to receive the analytics that is sent from the Miniapp.
+
+```swift
+extension ViewController: MiniAppMessageDelegate {
```
Comment (rleojoseph) [inline USERGUIDE.md:1159]:
```suggestion
extension ViewController: MAAnalayticsDelegate {
```

**Review by rleojoseph — APPROVED** [review]
Thanks a lot @GondiTilak . Great work man

**Pull request merged**